File: app/src/main/java/com/example/grocery/adapters/AdapterOrderUser.java
package com.example.grocery.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.grocery.activities.OrderDetailsUsersActivity;
import com.example.grocery.models.ModelOrderUser;
import com.example.phinh.grocery.R;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Calendar;

public class AdapterOrderUser extends RecyclerView.Adapter<AdapterOrderUser.HolderOrderUser> {

    private final Context context;
    private final ArrayList<ModelOrderUser> orderUserList;

    public AdapterOrderUser(Context context, ArrayList<ModelOrderUser> orderUserList) {
        this.context = context;
        this.orderUserList = orderUserList;
    }

    @NonNull
    @Override
    public HolderOrderUser onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //inflate layout
        View view = LayoutInflater.from(context).inflate(R.layout.row_order_user, parent , false);
        return new HolderOrderUser(view);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull HolderOrderUser holder, int position) {
        //get data
        ModelOrderUser modelOrderUser = orderUserList.get(position);
        final String orderId = modelOrderUser.getOrderId();
        String orderBy = modelOrderUser.getOrderBy();
        String orderCost = modelOrderUser.getOrderCost();
        String orderStatus = modelOrderUser.getOrderStatus();
        String orderTime = modelOrderUser.getOrderTime();
        final String orderTo = modelOrderUser.getOrderTo();

        //get shop info
        loadShopInfo(modelOrderUser, holder);

        //set data
        holder.amountTv.setText("Số lượng: $" +orderCost);
        holder.statusTv.setText(orderStatus);
        holder.orderIdTv.setText("ID đặt hàng: "+ orderId);
        //change order status text color
        switch (orderStatus) {
            case "Chưa duyệt":
                holder.statusTv.setTextColor(context.getResources().getColor(R.color.colorPrimary));
                break;
            case "Đã duyệt":
                holder.statusTv.setTextColor(context.getResources().getColor(R.color.colorGreen));
                break;
            case "Đã hủy":
                holder.statusTv.setTextColor(context.getResources().getColor(R.color.colorRed));
                break;
        }

        //convert timestamp to proper format
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(Long.parseLong(orderTime));
        String formatedDate = DateFormat.format("dd/MM/yyyy", calendar).toString(); //e.g. 16/06/2020

        holder.dateTv.setText(formatedDate);

        holder.itemView.setOnClickListener(v -> {
            //open order details, we need to keys there, orderId, orderTo
            Intent intent = new Intent(context, OrderDetailsUsersActivity.class);
            intent.putExtra("orderTo", orderTo);
            intent.putExtra("orderId", orderId);
            context.startActivity(intent);
        });
    }

    private void loadShopInfo(ModelOrderUser modelOrderUser, final HolderOrderUser holder) {
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Users");
        ref.child(modelOrderUser.getOrderTo())
                .addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        String shopName = ""+dataSnapshot.child("shopName").getValue();
                        holder.shopNameTv.setText(shopName);
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
    }

    @Override
    public int getItemCount() {
        return orderUserList.size();
    }

    //view holder class
    static class HolderOrderUser extends RecyclerView.ViewHolder{

        //views of layout
        private final TextView orderIdTv, dateTv, shopNameTv, amountTv, statusTv;

        public HolderOrderUser(@NonNull View itemView) {
            super(itemView);

            //init views of layout
            orderIdTv = itemView.findViewById(R.id.orderIdTv);
            dateTv = itemView.findViewById(R.id.dateTv);
            shopNameTv = itemView.findViewById(R.id.shopNameTv);
            amountTv = itemView.findViewById(R.id.amountTv);
            statusTv = itemView.findViewById(R.id.statusTv);
        }
    }
}

File: app/src/main/java/com/example/grocery/adapters/AdapterOrderUser.kt
package com.example.grocery.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.grocery.activities.OrderDetailsUsersActivity
import com.example.grocery.adapters.AdapterOrderUser.HolderOrderUser
import com.example.grocery.models.ModelOrderUser
import com.example.grocery.R
import com.example.grocery.activities.CurrencyFormatter
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.Calendar

class AdapterOrderUser(
    private val context: Context,
    private val orderUserList: ArrayList<ModelOrderUser?>?
) : RecyclerView.Adapter<HolderOrderUser>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HolderOrderUser {
        //inflate layout
        val view = LayoutInflater.from(context).inflate(R.layout.row_order_user, parent, false)
        return HolderOrderUser(view)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: HolderOrderUser, position: Int) {
        //get data
        val modelOrderUser = orderUserList!![position]
        val orderId = modelOrderUser?.orderId
        val orderBy = modelOrderUser?.orderBy
        val orderCost = modelOrderUser?.orderCost
        val orderStatus = modelOrderUser?.orderStatus
        val orderTime = modelOrderUser?.orderTime
        val orderTo = modelOrderUser?.orderTo

        //get shop info
        loadShopInfo(modelOrderUser!!, holder)
// Chuyển đổi orderCost từ String sang Double an toàn
        val orderCostDouble = orderCost?.toDoubleOrNull() ?: 0.0

        // Định dạng lại chuỗi hiển thị số tiền
        val formattedOrderCost = CurrencyFormatter.format(orderCostDouble)
        //set data
        holder.amountTv.text = "Số lượng: $formattedOrderCost"
        holder.statusTv.text = orderStatus
        holder.orderIdTv.text = "ID đặt hàng: $orderId"
        when (orderStatus) {
            "Chưa duyệt" -> holder.statusTv.setTextColor(context.resources.getColor(R.color.colorPrimary))
            "Đã duyệt" -> holder.statusTv.setTextColor(context.resources.getColor(R.color.colorGreen))
            "Đã hủy" -> holder.statusTv.setTextColor(context.resources.getColor(R.color.colorRed))
        }

        //convert timestamp to proper format
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = orderTime!!.toLong()
        val formatedDate = DateFormat.format("dd/MM/yyyy", calendar).toString() //e.g. 16/06/2020
        holder.dateTv.text = formatedDate
        holder.itemView.setOnClickListener { v: View? ->
            //open order details, we need to keys there, orderId, orderTo
            val intent = Intent(context, OrderDetailsUsersActivity::class.java)
            intent.putExtra("orderTo", orderTo)
            intent.putExtra("orderId", orderId)
            context.startActivity(intent)
        }
    }

    private fun loadShopInfo(modelOrderUser: ModelOrderUser, holder: HolderOrderUser) {
        val ref = FirebaseDatabase.getInstance().getReference("Users")
        ref.child(modelOrderUser.orderTo!!)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val shopName = "" + dataSnapshot.child("shopName").value
                    holder.shopNameTv.text = shopName
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    override fun getItemCount(): Int {
        return orderUserList!!.size
    }

    //view holder class
    class HolderOrderUser(itemView: View) : RecyclerView.ViewHolder(itemView) {
        //views of layout
        val orderIdTv: TextView
        val dateTv: TextView
        val shopNameTv: TextView
        val amountTv: TextView
        val statusTv: TextView

        init {

            //init views of layout
            orderIdTv = itemView.findViewById(R.id.orderIdTv)
            dateTv = itemView.findViewById(R.id.dateTv)
            shopNameTv = itemView.findViewById(R.id.shopNameTv)
            amountTv = itemView.findViewById(R.id.amountTv)
            statusTv = itemView.findViewById(R.id.statusTv)
        }
    }
}
